src/main/java/com/iota/iri/storage/rocksDB/RocksDBPPPImpl.java

```diff
@@ -0,0 +1,607 @@
+                    byte[] subarray2 = ArrayUtils.subarray(result, keyLoc + indexBytes.length + 1, result.length);
+                    if (subarray1.length > 0 && subarray2.length > 0) {
+                        buffer.put(delimiter);
+                        buffer.put(subarray2);
+                    }
+                    writeBatch.put(column, key.bytes(), buffer.array());
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds a key to an index through an atomic write batch
+     * @param writeBatch rocksDB atomic operation
+     * @param column index column
+     * @param key index key
+     * @param indexValue index value
+     * @throws Exception
+     */
+    @VisibleForTesting
+    void addToIndex(WriteBatch writeBatch, ColumnFamilyHandle column, Indexable key, Indexable indexValue)
+            throws Exception {
+        byte[] indexBytes = indexValue.bytes();
+        byte[] dbResult = db.get(column, key.bytes());
+        if (dbResult == null) {
+            dbResult = new byte[0];
+        }
+
+        int keyLoc = indexOf(dbResult, indexBytes);
+        if (keyLoc == -1) {// not found
+
+            ByteBuffer buffer = ByteBuffer
+                    .allocate(dbResult.length + indexBytes.length + (dbResult.length > 0 ? 1 : 0));// +1 delimiter
+                                                                                                   // length
+            buffer.put(dbResult);
+            if (dbResult.length > 0) {
+                // Means we add on the end of the current stream.
+                // To be compatible with Hashes.read(byte[])
+                buffer.put(delimiter);
+            }
+            buffer.put(indexBytes);
+            writeBatch.put(column, key.bytes(), buffer.array());
+        }
+    }
+
+    /**
+     * Byte matching variant of string.indexOf
+     * @param outerArray Array to search in
+     * @param smallerArray What to find
+     * @return
+     */
+    public int indexOf(byte[] outerArray, byte[] smallerArray) {
+        for (int i = 0; i < outerArray.length - smallerArray.length + 1; ++i) {
+            boolean found = true;
+            for (int j = 0; j < smallerArray.length; ++j) {
+                if (outerArray[i + j] != smallerArray[j]) {
+                    found = false;
+                    break;
+                }
+            }
+            if (found) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * @see PermanentPersistenceProvider#getTransaction(Hash)
+     */
+    @Override
+    public TransactionViewModel getTransaction(Hash index) throws Exception {
+        if (index == null) {
+            return null;
+        }
+        byte[] dbResult = db.get(columnMap.get(TRANSACTION_COLUMN), index.bytes());
+        if (dbResult != null && dbResult.length > 0) {
+            return new TransactionViewModel(TransactionViewModel.trits(dbResult), Hash.NULL_HASH);
+        }
+        return null;
```

Comment:
Making an empty static object would be fine with me as well but I think that is outside of the scope for this pull request. the function getTransaction is part of the PermanentPersistenceProvider and not of the PersistenceProvider.

I chose null because it is more explicit than an empty object. In the get function of the PersistenceProvider interface this is explicitly handled.

Reply:
It is a possible way to go about it.

However, the reason I preferred to go about an empty object is because then you have to remember to always do `null` checks. If you forget (like in `safeDeleteTransaction` :-P) you later have to go back and add it...

The thing is that probably the `PermanentPersistenceProvider` and `PersistenceProvider` will have to be similar in this sense eventually, so might as well get it right from the beginning :-)

I only asked you to add this empty object in `PermanentPersistenceProvider` and not in the old code. If you feel strongly against this then just know:
1. You will have to add lots of null checks ;-)
2. It will just make things messier later